### ISSUE TITLE
Numpy2.0 migration

### ISSUE THREAD

**Issue by smiet** [hiddenSymmetries/simsopt#427]
The container build is failing because the SPEC and VMEC2000 compilation step are failing. 

The migration guide to numpy2.0 is given [here](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide). 

I will work on getting SPEC up and running with numpy 2.0, @landreman can you have a look at VMEC? from [this link](https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-abi-handling) it looks like this is what should be added: 

```python
[build-system]
build-backend = ...
requires = [
    # Note for packagers: this constraint is specific to wheels
    # for PyPI; it is also supported to build against 1.xx still.
    # If you do so, please ensure to include a `numpy<2.0`
    # runtime requirement for those binary packages.
    "numpy>=2.0.0rc1",
    ...
]

[project]
dependencies = [
    "numpy>=1.23.5",
]
```

**Comment by landreman** [comment]
Hm, I tried that change in the vmec2000 repo, but there are various errors in the vmec2000 CI now. Looping in @mbkumar @andrewgiuliani @daringli @rogeriojorge in case anyone can help.

**Comment by smiet** [comment]
Now thinking we might want to temporarily force numpy<2.0 in all runners, there are other dependencies like f90wrap that use the deprecated oldest-supported-numpy that need to be patched, as well as that a new pyoculus needs to be uploaded to pypi as we resolve all the dependencies

**Comment by smiet** [comment]
Update: f90wrap needs to be adapted as well to make spec (and simsopt) run with numpy 2.0. Working with the upstream to merge a [pull-request](https://github.com/jameskermode/f90wrap/pull/194 ). 

In the meantime we can try to force the runners to use a 1.x version of numpy, as I tried in PR #429

Dependencies that still need to be patched:
- [ ] F90wrap @smiet
- [ ] SPEC @smiet 
- [ ] pyoculus @smiet 
- [ ] VMEC2000 @mkumbar @landreman @andrewgiuliani @daringli @rogeriojorge

**Comment by smiet** [comment]
nptyping does not seem to have any compatibility with numpy2.0, and when installed under 2.0 raises an exception upon import. 
Currently it's dependency is set to numpy<2.0, such that installing simsopt will downgrade numpy to 1.26.

Can we replace the dependency? @mbkumar, I see you added it, what does nptyping allow us to do that numpy.typing cannot provide?

**Comment by mbkumar** [comment]
Looks like development on nptying ceased last year. We should switch to `numpy.typing`. I'll work on it ASAP.

**Comment by smiet** [comment]
Tests are failing because of a few deprecated numpy uses. See the branch [ml/vmec_update_for_numpy_2](https://github.com/hiddenSymmetries/simsopt/tree/ml/vmec_update_for_numpy_2) and it's tests for progress. 

@florianwechsung the `randomgen` package is not updated for numpy 2.0, could np.random be used, or is it essential for the stochastic optimizations?

**Comment by smiet** [comment]
@mbkumar You need to use the main_off branch, this is up-to-date with remote, works with numpy2.0 and has an open pull request.

**Comment by smiet** [comment]
Sorry @mbkumar, the new branch also needs (for the time being) the `numpy2_fix` branch of SPEC. The [PR](https://github.com/PrincetonUniversity/SPEC/pull/199) is awaiting review. 

```
    - name: Download SPEC
      run: git clone -b numpy2_fix --depth=1 https://github.com/PrincetonUniversity/SPEC
```
Hopefully this PR will be resolved tomorrow.

**Comment by mbkumar** [comment]
@smiet Thanks for letting me know. I'll wait till the PRs are merged. Please ping me once the PRs are merged.

**Comment by smiet** [comment]
@mbkumar You are a contributor to SPEC, If you give [the PR](https://github.com/hiddenSymmetries/simsopt/issues/427#issuecomment-2189278030) a quick glance and agree with the changes, you can approve and merge.

**Comment by smiet** [comment]
@mbkumar The new version of SPEC is merged into master

**Comment by florianwechsung** [comment]
> Tests are failing because of a few deprecated numpy uses. See the branch [ml/vmec_update_for_numpy_2](https://github.com/hiddenSymmetries/simsopt/tree/ml/vmec_update_for_numpy_2) and it's tests for progress.
> 
> @florianwechsung the `randomgen` package is not updated for numpy 2.0, could np.random be used, or is it essential for the stochastic optimizations?

I vaguely recall using randomgen because the standard numpy.random does not satisfy certain independence conditions when running in parallel. I'm travelling this week but will have a look on the weekend. Please ping me again if I don't come back to you by Monday.

**Comment by mbkumar** [comment]
@florianwechsung,

Looks like most of the `randomgen` codebase has been merged into `numpy`. `numpy` is now supplying `PCG64` and its successor [PCG64DXSM](https://numpy.org/doc/stable/reference/random/bit_generators/pcg64dxsm.html#numpy.random.PCG64DXSM). The successor seems to solve the parallel issue. 

If you don't have time, I can take the lead in replacing`randomgen.PCG64` with `numpy.random.PCG64DXSM`.

**Comment by smiet** [comment]
Closing as randomgen has been updated now. Nevertheless opened PR  #517 to move to numpy and decrease our depencency.
